ext/intl/rangeformatter/rangeformatter_class.cpp

```diff
@@ -0,0 +1,215 @@
+/*
+   +----------------------------------------------------------------------+
+   | This source file is subject to version 3.01 of the PHP license,      |
+   | that is bundled with this package in the file LICENSE, and is        |
+   | available through the world-wide-web at the following url:           |
+   | https://www.php.net/license/3_01.txt                                 |
+   | If you did not receive a copy of the PHP license and are unable to   |
+   | obtain it through the world-wide-web, please send a note to          |
+   | license@php.net so we can mail you a copy immediately.               |
+   +----------------------------------------------------------------------+
+   | Authors: Bogdan Ungureanu <bogdanungureanu21@gmail.com>              |
+   +----------------------------------------------------------------------+
+*/
+
+#include <unicode/numberrangeformatter.h>
+#include <unicode/unumberrangeformatter.h>
+#include <unicode/numberformatter.h>
+#include <unicode/unistr.h>
+#include "../intl_convertcpp.h"
+
+extern "C" {
+    #include "php.h"
+    #include "zend_API.h"
+    #include "../intl_common.h"
+    #include "../intl_error.h"
+    #include "../php_intl.h"
+    #include "../intl_data.h"
+    #include "rangeformatter_arginfo.h"
+    #include "rangeformatter_class.h"
+    #include "intl_convert.h"
+}
+
+using icu::number::NumberRangeFormatter;
+using icu::number::NumberFormatter;
+using icu::number::UnlocalizedNumberFormatter;
+using icu::number::LocalizedNumberRangeFormatter;
+using icu::UnicodeString;
+using icu::MeasureUnit;
+
+static zend_object_handlers rangeformatter_handlers;
+zend_class_entry *class_entry_IntlNumberRangeFormatter;
+
+zend_object *IntlNumberRangeFormatter_object_create(zend_class_entry *ce)
+{
+    IntlNumberRangeFormatter_object* intern;
+
+    intern = reinterpret_cast<IntlNumberRangeFormatter_object*>(zend_object_alloc(sizeof(IntlNumberRangeFormatter_object), ce));
+    zend_object_std_init(&intern->zo, ce);
+    object_properties_init(&intern->zo, ce);
+
+    // Initialize rangeformatter_data structure
+    intl_error_init(&intern->nrf_data.error);
+    intern->nrf_data.unumrf = nullptr;
+
+    intern->zo.handlers = &rangeformatter_handlers;
+
+    return &intern->zo;
+}
+
+U_CFUNC PHP_METHOD(IntlNumberRangeFormatter, __construct)
+{
+    ZEND_PARSE_PARAMETERS_NONE();
+    zend_throw_error(NULL, "Cannot directly construct %s, use createFromSkeleton method instead", ZSTR_VAL(Z_OBJCE_P(ZEND_THIS)->name));
+}
+
+U_CFUNC PHP_METHOD(IntlNumberRangeFormatter, createFromSkeleton)
+{
+#if U_ICU_VERSION_MAJOR_NUM < 63
```

Comment:
? So how will this even compile then on older versions? Shouldn't this entire class be compiled conditionally instead of adding exception code at the top?

Reply:
Yes, that's correct, it would compile to other versions. I think it's more developer friendly to know why the class is not available with their configuration - doing it like this, the developer knows that the class is not available because of the ICU constraint. Otherwise, they'll get a fatal for a non-existent class which would feel confusing imo.

It was previously discussed [here](https://github.com/php/php-src/pull/19232#discussion_r2231650284) too. :)

Reply:
Normally we only make classes available conditionally.
Anyway, besides that, the main concern is not answered: how will this compile on versions below ICU 62? You still compile the code below the `#if` block which presumably is only compilable on ICU 62 and higher?

Reply:
Done. I've ran the tests with a quick and dirty [docker container](https://gist.github.com/BogdanUngureanu/26976b10454b5dcb388fe2adacb1d40c) running icu62 and the failure test passed.